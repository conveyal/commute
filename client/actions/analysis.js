import moment from 'moment'
import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addAnalysis = createAction('add analysis')
export const createAnalysis = (newAnalysis, organizationId) => {
  newAnalysis.id = uuid.v4()
  newAnalysis.lastRunDateTime = moment().unix()
  const commuters = newAnalysis.commuters
  delete newAnalysis.commuters
  return [
    addAnalysis({ analysis: newAnalysis, organizationId }),
    push(`/organizations/${organizationId}/analysis/${newAnalysis.id}`),
    makeMockTrips(newAnalysis.id, organizationId, commuters)
  ] // TODO save to server
}

const deleteLocally = createAction('delete analysis')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/analysis/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteAnalysis = (id, organizationId) => [
  deleteLocally({ analysisId: id, organizationId }),
  push(`/organizations/${organizationId}`)
]

const makeMockTrips = (analysisId, organizationId, commuters) => {
  const trips = []
  const modes = ['bike', 'car', 'transit', 'walk']
  const randRange = () => Math.random() * 0.4 + 0.8
  commuters.forEach((commuter) => {
    const curTrip = {
      commuterId: commuter.id
    }
    const tripMetricBase = Math.random()
    let bestValue = 9999999
    let bestMode
    modes.forEach((mode) => {
      // generate random trips
      const modePctMultiplier = randRange()
      const cost = tripMetricBase * 34.7 * modePctMultiplier * randRange()
      const distance = tripMetricBase * 69 * modePctMultiplier * randRange()
      const time = tripMetricBase * 7500 * modePctMultiplier * randRange()
      curTrip[mode] = {cost, distance, time}
      const tripValue = cost + time * 0.01
      if (tripValue < bestValue) bestMode = mode
    })
    curTrip.mostLikely = Object.assign({ mode: bestMode }, curTrip[bestMode])
    trips.push(curTrip)
  })
  return sendMockTrips({analysisId, organizationId, trips})
}
const sendMockTrips = createAction('receive mock calculated trips')
