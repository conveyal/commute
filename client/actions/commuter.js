import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addCommuter = createAction('add commuter')
export const createCommuter = ({commuter, groupId, organizationId}) => {
  commuter.id = uuid.v4()
  return [
    addCommuter({commuter, groupId, organizationId}),
    push(`/organizations/${organizationId}/groups/${groupId}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete commuter')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/commuter/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteCommuter = ({commuterId, groupId, organizationId}) => [
  deleteLocally({commuterId, groupId, organizationId}),
  push(`/organizations/${organizationId}/groups/${groupId}`)
]

const updateLocally = createAction('update commuter')
/* const updateOnServer = (id) =>
  serverAction({
    url: `/api/commuter/${id}`,
    options: {
      method: 'update'
    }
  }) */ // TODO update on server
export const updateCommuter = ({commuter, groupId, organizationId}) => [
  updateLocally({commuter, groupId, organizationId}),
  push(`/organizations/${organizationId}/groups/${groupId}`)
]
