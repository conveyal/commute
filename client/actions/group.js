import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addGroup = createAction('add group')
const appendCommuters = createAction('append commuters')
export const appendToGroup = (newCommuters, groupId, organizationId) => {
  return [
    appendCommuters({ commuters: newCommuters, groupId, organizationId }),
    push(`/organizations/${organizationId}/groups/${groupId}`)
  ] // TODO save to server
}
export const createGroup = (newGroup, organizationId) => {
  newGroup.id = uuid.v4()
  newGroup.allAddressesGeocoded = false
  return [
    addGroup({ group: newGroup, organizationId }),
    push(`/organizations/${organizationId}/groups/${newGroup.id}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete group')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/group/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteGroup = (id, organizationId) => [
  deleteLocally({ groupId: id, organizationId }),
  push(`/organizations/${organizationId}`)
]
