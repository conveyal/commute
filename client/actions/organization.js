import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addLocally = createAction('add organization')
export const createOrganization = (newOrganization) => {
  newOrganization.id = uuid.v4()
  newOrganization.analyses = []
  newOrganization.analysesById = {}
  newOrganization.sites = []
  newOrganization.sitesById = {}
  newOrganization.groups = []
  newOrganization.groupsById = {}
  return [
    addLocally(newOrganization),
    push(`/organizations/${newOrganization.id}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete organization')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/organization/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteOrganization = (id) => [deleteLocally(id), push('/')]

export const updateOrganization = (organization) => [
  setLocally(organization),
  push(`/organizations/${organization.id}`)
]
/* const updateOnServer = (organization) =>
  serverAction({
    url: `/api/organization/${organization.id}`,
    options: {
      body: JSON.stringify(organization),
      method: 'put'
    }
  }) */  // TODO: update on server

export const setLocally = createAction('set organization')
