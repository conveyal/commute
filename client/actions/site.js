import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addSite = createAction('add site')
export const createSite = (newSite, organizationId) => {
  newSite.id = uuid.v4()
  return [
    addSite({ organizationId, site: newSite }),
    push(`/organizations/${organizationId}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete site')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/site/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteSite = (id, organizationId) => [
  deleteLocally({ siteId: id, organizationId }),
  push(`/organizations/${organizationId}`)
]

const updateLocally = createAction('update site')
/* const updateOnServer = (id) =>
  serverAction({
    url: `/api/site/${id}`,
    options: {
      method: 'update'
    }
  }) */ // TODO update on server
export const updateSite = (site, organizationId) => [
  updateLocally({ organizationId, site }),
  push(`/organizations/${organizationId}`)
]
