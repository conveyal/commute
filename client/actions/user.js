import {createAction} from 'redux-actions'

// login / logout
export const login = createAction('log in')
export const logout = createAction('log out')
