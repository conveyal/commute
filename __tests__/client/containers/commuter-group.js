/* global describe, expect, it */

import {mount} from 'enzyme'
import {mountToJson} from 'enzyme-to-json'
import pretty from 'pretty'
import React from 'react'
import {Provider} from 'react-redux'

import {makeGenericModelActionsExpectations} from '../../test-utils/actions'
import {makeMockStore, mockCommuter, mockStores} from '../../test-utils/mock-data.js'
import '../../test-utils/mock-leaflet'

import CommuterGroup from '../../../client/containers/commuter-group'

const commuterExpectations = makeGenericModelActionsExpectations({
  pluralName: 'commuters',
  singularName: 'commuter'
})

describe('Container > CommuterGroup', () => {
  it('Commuter Group View loads', () => {
    const mockStore = makeMockStore(mockStores.withAnalysisRun)

    // mount component
    const tree = mount(
      <Provider store={mockStore}>
        <CommuterGroup
          params={{groupId: 'group-2'}}
          />
      </Provider>
      , {
        attachTo: document.getElementById('test')
      }
    )

    expect(mountToJson(tree.find('.group-header'))).toMatchSnapshot()
    expect(pretty(tree.find('.group-content').html())).toMatchSnapshot()

    // TODO: cluster should be added initially
  })

  /* it('Edit commuter group name', () => {
    throw new Error('unimplemented')
  }) */

  it('Delete Commuter', () => {
    const mockStore = makeMockStore(mockStores.withAnalysisRun)
    window.confirm = () => true

    // Given a logged-in user is viewing the commuter group view
    // mount component
    const tree = mount(
      <Provider store={mockStore}>
        <CommuterGroup
          params={{groupId: 'group-2'}}
          />
      </Provider>
      , {
        attachTo: document.getElementById('test')
      }
    )

    // clear all loading actions
    mockStore.clearActions()

    // When the user clicks the delete button for an existing site
    // And the user confirms the Confirm Deletion dialog
    const deleteButton = tree.find('table').at(1).find('button').last()
    deleteButton.simulate('click')

    commuterExpectations.expectDeleteAction({
      action: mockStore.getActions()[0],
      entity: mockCommuter
    })
  })

  /* it('Show Commuter in Ridematch', () => {
    throw new Error('unimplemented')
  }) */
})
