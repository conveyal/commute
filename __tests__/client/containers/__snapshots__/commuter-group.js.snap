exports[`Container > CommuterGroup Commuter Group View loads 1`] = `
<div
  className="group-header col-xs-12">
  <h3>
    <span>
      Mock Group
    </span>
    <ButtonLink
      className="pull-right"
      to="/organization/organization-2">
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="default"
        className="pull-right"
        disabled={false}
        onClick={[Function]}
        to="/organization/organization-2">
        <button
          className="pull-right btn btn-default"
          disabled={false}
          onClick={[Function]}
          to="/organization/organization-2"
          type="button">
          <Icon
            type="arrow-left">
            <i
              className="fa fa-arrow-left fa-fw " />
          </Icon>
          <span>
            Back
          </span>
        </button>
      </Button>
    </ButtonLink>
  </h3>
  <ButtonGroup
    block={false}
    bsClass="btn-group"
    justified={false}
    vertical={false}>
    <div
      className="btn-group">
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="warning"
        disabled={false}
        onClick={[Function]}>
        <button
          className="btn btn-warning"
          disabled={false}
          onClick={[Function]}
          type="button">
          <Icon
            type="pencil">
            <i
              className="fa fa-pencil fa-fw " />
          </Icon>
          <span>
            Edit Name
          </span>
        </button>
      </Button>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="danger"
        disabled={false}
        onClick={[Function]}>
        <button
          className="btn btn-danger"
          disabled={false}
          onClick={[Function]}
          type="button">
          Delete Group
        </button>
      </Button>
    </div>
  </ButtonGroup>
</div>
`;

exports[`Container > CommuterGroup Commuter Group View loads 2`] = `
"<div class=\"group-content row\">
  <div class=\"col-xs-12\">
    <h3>Commuters</h3>
    <div class=\"btn-group\">
      <button to=\"/group/group-2/commuter/create\" type=\"button\" class=\"btn btn-info\">
        <i class=\"fa fa-plus fa-fw \"></i><span>Create New Commuter</span> </button>
      <button to=\"/group/group-2/add\" type=\"button\" class=\"btn btn-warning\">Add Commuters in Bulk</button>
    </div>
    <div class=\"react-bs-table-container\">
      <div class=\"react-bs-table\" style=\"height: 100%;\">
        <div class=\"react-bs-container-header table-header-wrapper\">
          <table class=\"table table-hover table-bordered\">
            <colgroup>
              <col class=\"\" style=\"display: none; min-width: NaNpx;\">
                <col class=\"\" style=\"min-width: NaNpx;\">
                  <col class=\"\" style=\"min-width: NaNpx;\">
                    <col class=\"\" style=\"min-width: NaNpx;\">
                      <col class=\"\" style=\"min-width: NaNpx;\">
                        <col class=\"\" style=\"min-width: NaNpx;\">
                          <col class=\"\" style=\"min-width: NaNpx;\">
            </colgroup>
            <thead>
              <tr>
                <th class=\"\" data-is-only-head=\"false\" style=\"text-align: left; display: none;\" data-field=\"_id\">
                  <div></div>
                </th>
                <th class=\"\" data-is-only-head=\"false\" title=\"Name\" style=\"text-align: left;\" data-field=\"name\">

                  <!-- react-text: 41 -->Name<!-- /react-text -->
                  <div></div>
                </th>
                <th class=\"\" data-is-only-head=\"false\" title=\"Original Address\" style=\"text-align: left;\" data-field=\"original_address\">

                  <!-- react-text: 44 -->Original Address<!-- /react-text -->
                  <div></div>
                </th>
                <th class=\"\" data-is-only-head=\"false\" title=\"Geocoded Address\" style=\"text-align: left;\" data-field=\"address\">

                  <!-- react-text: 47 -->Geocoded Address<!-- /react-text -->
                  <div></div>
                </th>
                <th class=\"\" data-is-only-head=\"false\" title=\"Geocode Confidence\" style=\"text-align: left;\" data-field=\"undefined\">

                  <!-- react-text: 50 -->Geocode Confidence<!-- /react-text -->
                  <div></div>
                </th>
                <th class=\"\" data-is-only-head=\"false\" title=\"Email\" style=\"text-align: left;\" data-field=\"email\">

                  <!-- react-text: 53 -->Email<!-- /react-text -->
                  <div></div>
                </th>
                <th class=\"\" data-is-only-head=\"false\" title=\"Tools\" style=\"text-align: left;\" data-field=\"undefined\">

                  <!-- react-text: 56 -->Tools<!-- /react-text -->
                  <div></div>
                </th>
              </tr>
            </thead>
          </table>
        </div>
        <div class=\"react-bs-container-body\" style=\"height: 100%;\">
          <table class=\"table table-bordered\">
            <colgroup>
              <col class=\"\" style=\"display: none;\">
                <col class=\"\">
                  <col class=\"\">
                    <col class=\"\">
                      <col class=\"\">
                        <col class=\"\">
                          <col class=\"\">
            </colgroup>
            <tbody>
              <tr class=\"\">
                <td class=\"\" style=\"text-align: left; display: none;\">commuter-2</td>
                <td class=\"\" style=\"text-align: left;\">Luke Warm</td>
                <td class=\"\" style=\"text-align: left;\"></td>
                <td class=\"\" style=\"text-align: left;\">4321 XYZ Boulevard</td>
                <td class=\"\" style=\"text-align: left;\">
                  <div>77 %</div>
                </td>
                <td class=\"\" style=\"text-align: left;\">luke@warm.cold</td>
                <td class=\"\" style=\"text-align: left;\">
                  <div class=\"btn-group\">
                    <button to=\"/commuter/commuter-2/edit\" type=\"button\" class=\"btn btn-warning\">Edit</button>
                    <button type=\"button\" class=\"btn btn-danger\">Delete</button>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>"
`;

exports[`Container > CommuterGroup Delete Commuter 1`] = `
Object {
  "payload": Object {
    "args": Array [
      "/group/group-2",
    ],
    "method": "push",
  },
  "type": "@@router/CALL_HISTORY_METHOD",
}
`;
